### ISSUE TITLE
shadows-playservices causes bytecode VerifyError

### ISSUE THREAD

**Issue by SamMaier** [robolectric/robolectric#7269]
### Description
Calling the constructor of [play-services-basement's](https://mvnrepository.com/artifact/com.google.android.gms/play-services-basement) [ApiException](https://developers.google.com/android/reference/com/google/android/gms/common/api/ApiException#ApiException(com.google.android.gms.common.api.Status)), when using robolectric's [shadows-playservics](https://mvnrepository.com/artifact/org.robolectric/shadows-playservices), causes a VerifyError:

```
java.lang.VerifyError: Bad local variable type
Exception Details:
  Location:
    com/google/android/gms/common/api/ApiException.<init>(Lcom/google/android/gms/common/api/Status;)V @1: aload
  Reason:
    Type top (current frame, locals[5]) is not assignable to reference type
  Current Frame:
    bci: @1
    flags: { flagThisUninit }
    locals: { uninitializedThis, 'com/google/android/gms/common/api/Status' }
    stack: { uninitializedThis }
  Bytecode:
    0000000: 2a19 05b6 0041 b700 442a b600 482a 2bba
    0000010: 0053 0000 b1                           

	at smaier.SimpleRobolectricTest.testApiException(SimpleRobolectricTest.java:21)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:591)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:274)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:88)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:830)
```

### Steps to Reproduce
Just call `ApiException.<init>` in any robolectric test with shadows-playservices and the newest version (18.0.1) of play-services-basement.

### Robolectric & Android Version
This reproduced on robolectric 4.7.3 and some random 4.3 version - I don't think this is a recent bug in robolectric. This does not happen with versions of play-services-basement 17.5.0 and earlier, but does happen with 17.6.0 and later. However, if you don't use shadows-playservices, this does not happen - so I do believe this is an issue with shadows-playservices and how it interacts with the newer version of play-services-basement. 

### Link to a public git repo demonstrating the problem:
https://github.com/SamMaier/basement_robolectric_failure

To use this, you can just run `build_and_run_robolectric_test.sh`. If you want to play with different versions of deps, edit the `pom.xml` file and run `rm -f target/dependency/* && mvn dependency:copy-dependencies`. If you remove shadows-playservices from the `pom.xml` file and update the deps, you will notice that the test will now pass.

**Comment by hoisie** [comment]
Interesting edge case. I have no idea how this class was compiled. There seems to be complex logic in the ApiException constructor that happens before invoke `super`, which is causing issues for constructor instrumentation.

com.google.android.gms.common.api.ApiException before instrumentation:
```
$ javap -p -c 'com.google.android.gms.common.api.ApiException-orig-8.class'
Compiled from "com.google.android.gms:play-services-basement@@18.0.1"
public class com.google.android.gms.common.api.ApiException extends java.lang.Exception {
  protected final com.google.android.gms.common.api.Status mStatus;

  public int getStatusCode();
    Code:
       0: aload_0
       1: getfield      #13                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
       4: invokevirtual #17                 // Method com/google/android/gms/common/api/Status.getStatusCode:()I
       7: ireturn

  public com.google.android.gms.common.api.Status getStatus();
    Code:
       0: aload_0
       1: getfield      #13                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
       4: areturn

  public java.lang.String getStatusMessage();
    Code:
       0: aload_0
       1: getfield      #13                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
       4: invokevirtual #26                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
       7: areturn

  public com.google.android.gms.common.api.ApiException(com.google.android.gms.common.api.Status);
    Code:
       0: aload_1
       1: invokevirtual #17                 // Method com/google/android/gms/common/api/Status.getStatusCode:()I
       4: istore_3
       5: aload_1
       6: invokevirtual #26                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
       9: ifnull        82
      12: aload_1
      13: invokevirtual #26                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
      16: astore_2
      17: aload_2
      18: invokestatic  #34                 // Method java/lang/String.valueOf:(Ljava/lang/Object;)Ljava/lang/String;
      21: astore        4
      23: aload         4
      25: invokevirtual #37                 // Method java/lang/String.length:()I
      28: istore        4
      30: new           #39                 // class java/lang/StringBuilder
      33: astore        5
      35: aload         5
      37: iload         4
      39: bipush        13
      41: iadd
      42: invokespecial #42                 // Method java/lang/StringBuilder."<init>":(I)V
      45: aload         5
      47: iload_3
      48: invokevirtual #46                 // Method java/lang/StringBuilder.append:(I)Ljava/lang/StringBuilder;
      51: pop
      52: aload         5
      54: ldc           #48                 // String :
      56: invokevirtual #51                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      59: pop
      60: aload         5
      62: aload_2
      63: invokevirtual #51                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      66: pop
      67: aload_0
      68: aload         5
      70: invokevirtual #54                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      73: invokespecial #57                 // Method java/lang/Exception."<init>":(Ljava/lang/String;)V
      76: aload_0
      77: aload_1
      78: putfield      #13                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
      81: return
      82: ldc           #59                 // String
      84: astore_2
      85: goto          17
}
```

After instrumentation:
```
$ javap -p -c '/tmp/com.google.android.gms.common.api.ApiException-robo-instrumented-8.class'
Compiled from "com.google.android.gms:play-services-basement@@18.0.1"
public class com.google.android.gms.common.api.ApiException extends java.lang.Exception implements org.robolectric.internal.bytecode.ShadowedObject {
  public java.lang.Object __robo_data__;

  protected com.google.android.gms.common.api.Status mStatus;

  private final int $$robo$$com_google_android_gms_common_api_ApiException$getStatusCode();
    Code:
       0: aload_0
       1: getfield      #17                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
       4: invokevirtual #22                 // Method com/google/android/gms/common/api/Status.getStatusCode:()I
       7: ireturn

  private final com.google.android.gms.common.api.Status $$robo$$com_google_android_gms_common_api_ApiException$getStatus();
    Code:
       0: aload_0
       1: getfield      #17                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
       4: areturn

  private final java.lang.String $$robo$$com_google_android_gms_common_api_ApiException$getStatusMessage();
    Code:
       0: aload_0
       1: getfield      #17                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
       4: invokevirtual #32                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
       7: areturn

  private void $$robo$$com_google_android_gms_common_api_ApiException$__constructor__(com.google.android.gms.common.api.Status);
    Code:
       0: aload_1
       1: invokevirtual #22                 // Method com/google/android/gms/common/api/Status.getStatusCode:()I
       4: istore_3
       5: aload_1
       6: invokevirtual #32                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
       9: ifnull        73
      12: aload_1
      13: invokevirtual #32                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
      16: astore_2
      17: aload_2
      18: invokestatic  #40                 // Method java/lang/String.valueOf:(Ljava/lang/Object;)Ljava/lang/String;
      21: astore        4
      23: aload         4
      25: invokevirtual #43                 // Method java/lang/String.length:()I
      28: istore        4
      30: new           #45                 // class java/lang/StringBuilder
      33: astore        5
      35: aload         5
      37: iload         4
      39: bipush        13
      41: iadd
      42: invokespecial #49                 // Method java/lang/StringBuilder."<init>":(I)V
      45: aload         5
      47: iload_3
      48: invokevirtual #53                 // Method java/lang/StringBuilder.append:(I)Ljava/lang/StringBuilder;
      51: pop
      52: aload         5
      54: ldc           #55                 // String :
      56: invokevirtual #58                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      59: pop
      60: aload         5
      62: aload_2
      63: invokevirtual #58                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      66: pop
      67: aload_0
      68: aload_1
      69: putfield      #17                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
      72: return
      73: ldc           #60                 // String
      75: astore_2
      76: goto          17

  public int getStatusCode();
    Code:
       0: aload_0
       1: invokedynamic #74,  0             // InvokeDynamic #0:getStatusCode:(Lcom/google/android/gms/common/api/ApiException;)I
       6: ireturn

  public com.google.android.gms.common.api.Status getStatus();
    Code:
       0: aload_0
       1: invokedynamic #81,  0             // InvokeDynamic #1:getStatus:(Lcom/google/android/gms/common/api/ApiException;)Lcom/google/android/gms/common/api/Status;
       6: areturn

  public java.lang.String getStatusMessage();
    Code:
       0: aload_0
       1: invokedynamic #87,  0             // InvokeDynamic #2:getStatusMessage:(Lcom/google/android/gms/common/api/ApiException;)Ljava/lang/String;
       6: areturn

  private void __constructor__(com.google.android.gms.common.api.Status);
    Code:
       0: aload_0
       1: aload_1
       2: invokespecial #90                 // Method $$robo$$com_google_android_gms_common_api_ApiException$__constructor__:(Lcom/google/android/gms/common/api/Status;)V
       5: return

  public com.google.android.gms.common.api.ApiException(com.google.android.gms.common.api.Status);
    Code:
       0: aload_0
       1: aload         5
       3: invokevirtual #93                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
       6: invokespecial #96                 // Method java/lang/Exception."<init>":(Ljava/lang/String;)V
       9: aload_0
      10: invokevirtual #100                // Method $$robo$init:()V
      13: aload_0
      14: aload_1
      15: invokedynamic #104,  0            // InvokeDynamic #3:__constructor__:(Lcom/google/android/gms/common/api/ApiException;Lcom/google/android/gms/common/api/Status;)V
      20: return

  public com.google.android.gms.common.api.ApiException();
    Code:
       0: aload_0
       1: invokespecial #106                // Method java/lang/Exception."<init>":()V
       4: aload_0
       5: invokevirtual #100                // Method $$robo$init:()V
       8: return

  protected void $$robo$init();
    Code:
       0: aload_0
       1: getfield      #108                // Field __robo_data__:Ljava/lang/Object;
       4: ifnonnull     17
       7: aload_0
       8: aload_0
       9: invokedynamic #117,  0            // InvokeDynamic #4:initializing:(Lcom/google/android/gms/common/api/ApiException;)Ljava/lang/Object;
      14: putfield      #108                // Field __robo_data__:Ljava/lang/Object;
      17: return

  public java.lang.Object $$robo$getData();
    Code:
       0: aload_0
       1: getfield      #108                // Field __robo_data__:Ljava/lang/Object;
       4: areturn
}
```

The `1: aload         5` is clearly not right in the instrumented version.

**Comment by hoisie** [comment]
It seems like in Robolectric's instrumentation, there is an assumption that constructors are structured like this:

```
aload_0
... set up stack to call super(...) or this(...)
invokespecial (call super(...) or this(...)
.. all the stuff that happens after the call to super(...) or this(...)
```

Robolectric splits this up into two methods: a constructor that has `aload_0` upto and including `invokespecial`, and then a `__constructor__` method with the other stuff.

It seems that there is bytecode in the wild that does not have this format. This bytecode seems to defer the aload_0 until it is necessary.


So maybe instead of starting at `aload_0`, we should just start earlier and capture everything up until the call to `this(...)` or `super(...)`.

**Comment by hoisie** [comment]
I tracked this down a bit and unfortunately there's not really much we can do about this.

The issue is that ApiException from `play-services-basement` is run through dex/appreduce.

Before dex/appreduce, the constructor is something that Robolectric can process:

```
  public com.google.android.gms.common.api.ApiException(com.google.android.gms.common.api.Status);
    Code:
       0: aload_0
       1: aload_1
       2: invokevirtual #1                  // Method com/google/android/gms/common/api/Status.getStatusCode:()I
       5: aload_1
       6: invokevirtual #2                  // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
       9: ifnull        19
      12: aload_1
      13: invokevirtual #2                  // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
      16: goto          21
      19: ldc           #3                  // String
      21: invokedynamic #4,  0              // InvokeDynamic #0:makeConcatWithConstants:(ILjava/lang/String;)Ljava/lang/String;
      26: invokespecial #5                  // Method java/lang/Exception."<init>":(Ljava/lang/String;)V
      29: aload_0
      30: aload_1
      31: putfield      #6                  // Field mStatus:Lcom/google/android/gms/common/api/Status;
      34: return
```

After, it gets converted to this:

```
  public com.google.android.gms.common.api.ApiException(com.google.android.gms.common.api.Status);
    Code:
       0: aload_1
       1: invokevirtual #17                 // Method com/google/android/gms/common/api/Status.getStatusCode:()I
       4: istore_3
       5: aload_1
       6: invokevirtual #26                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
       9: ifnull        82
      12: aload_1
      13: invokevirtual #26                 // Method com/google/android/gms/common/api/Status.getStatusMessage:()Ljava/lang/String;
      16: astore_2
      17: aload_2
      18: invokestatic  #34                 // Method java/lang/String.valueOf:(Ljava/lang/Object;)Ljava/lang/String;
      21: astore        4
      23: aload         4
      25: invokevirtual #37                 // Method java/lang/String.length:()I
      28: istore        4
      30: new           #39                 // class java/lang/StringBuilder
      33: astore        5
      35: aload         5
      37: iload         4
      39: bipush        13
      41: iadd
      42: invokespecial #42                 // Method java/lang/StringBuilder."<init>":(I)V
      45: aload         5
      47: iload_3
      48: invokevirtual #46                 // Method java/lang/StringBuilder.append:(I)Ljava/lang/StringBuilder;
      51: pop
      52: aload         5
      54: ldc           #48                 // String :
      56: invokevirtual #51                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      59: pop
      60: aload         5
      62: aload_2
      63: invokevirtual #51                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      66: pop
      67: aload_0
      68: aload         5
      70: invokevirtual #54                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      73: invokespecial #57                 // Method java/lang/Exception."<init>":(Ljava/lang/String;)V
      76: aload_0
      77: aload_1
      78: putfield      #13                 // Field mStatus:Lcom/google/android/gms/common/api/Status;
      81: return
      82: ldc           #59                 // String
      84: astore_2
      85: goto          17
```

I tried various things, but couldn't get anything working. The complex parts are the instructions at the end (ldc/astore_2/goto) which refers back to the ones in the beginning.

In the meantime, I would recommend disabling this class from instrumentation:

```java
 public static class CustomRobolectricTestRunner extends RobolectricTestRunner {
   public CustomRobolectricTestRunner(Class<?> testClass) throws InitializationError {
    super(testClass);
  }

  @Override
  protected InstrumentationConfiguration createClassLoaderConfig(FrameworkMethod method) {
    return new InstrumentationConfiguration.Builder(super.createClassLoaderConfig(method))
        .doNotInstrumentClass("com.google.android.gms.common.api.ApiException")
        .build();
  }
}
```

Unfortunately at this point, we will need to deprecate and eventually remove shadows-playservices, as it seems like it's been broken for a while.

**Comment by agrieve** [comment]
Is there a way to apply this fix when using a ParameterizedRobolectricTestRunner? The class is final, so we can't override that method on it.

**Comment by hoisie** [comment]
@agrieve To do that you may need to fork ParameterizedRobolectricTestRunner (i.e copy the code and call it something like `MyParameterizedRobolectricTestRunner`), and then update MyParameterizedRobolectricTestRunner to use a RobolectricTestRunner subclass that has the relevant methods subclassed.

I don't think there's a way to do it via `@Config` or robolectric.properties :( Maybe we should add it.

**Comment by hoisie** [comment]
I may have some spare cycles to work on this over the next few weeks.

TLDR is that constructor instrumentation is very rarely required, and often leads to problems because it is easy to misuse it (i.e. double-calling constructors), so we may be able to detect patterns like this, and just skip the constructor instrumentation if necessary.

A true fix would involve having much more sophisticated splitting logic for the constructors that would require something akin to a graph traversal and reordering bytecode operations. 

Maybe I will also file a bug against GMSCore to see if they are intentionally shipping desugared bytecode.

**Comment by agrieve** [comment]
In our case, it turned out to be straight-forward to just not use the shadows, so we are no long hampered by this.
